### ISSUE TITLE
report an error when an attribute is not initialized

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ty#2954]
```py
class A:
    foo: int # no error

A().foo # no error, crashes at runtime
```

https://play.ty.dev/0579ef17-af4b-4d5b-bc94-757e4bb283e6

basically ty equivalent to pyright's `reportUninitializedInstanceVariable` rule

**Comment by AlexWaygood** [comment]
It's possible we could add such a rule at some point, but I think it's very low priority and will always be disabled by default, even if it is added at some point. It's very hard to do this kind of analysis in a foolproof way in Python without many false positives, because there are so many situations where it's expected that an attribute will be dynamically assigned in a nonstandard way. Pyright's version of this rule has [been effectively unmaintained](https://github.com/microsoft/pyright/issues/4687#issuecomment-1443998532) for a long time, and is not enabled even if strict mode has been selected.

**Comment by DetachHead** [comment]
In my experience, while that rule has some false positives, it's caught enough mistakes that I think it's worth having. I also think with the right heuristics it can be fine tuned to be much more reliable than the pyright version. For example https://docs.basedpyright.com/latest/benefits-over-pyright/improved-reportUninitializedInstanceVariable

**Comment by AlexWaygood** [comment]
It's definitely something we could consider adding at some point, and I'm sure you're right that it's possible to do better than pyright here
